app/src/main/java/ciserver/Pipeline.java

```diff
@@ -8,65 +11,64 @@ enum PipelineStatus {
 	InProgress
 }
 
+interface PipelineComponent {
+	public PipelineStatus execute(String pipelineDir, PushEvent event);
+}
+
 /**
  * A PipelineInstance is responsible for the whole CI-pipeline
  * for a specific commit.
  */
 class Pipeline {
-	String pipelineDir;
-	Commit commit;
+	final String pipelineDir;
+	final PushEvent event;
 	PipelineStatus status;
 
-	Pipeline(Commit commit, String pipelineDir) {
-		this.commit = commit;
+	/*
+	 * Continue with PipelineLinter, PipelineCompiler
+	 * Reason for this abstraction is easier testability,
+	 * As one can individually test the pull, lint, compile functionality,
+	 * without having a bunch of public methods in the Pipeline class.
+	 */
+	ArrayList<PipelineComponent> components = new ArrayList<>();
+
+	Pipeline(PushEvent event, String pipelineDir) {
+		this.event = event;
 		this.pipelineDir = pipelineDir;
-		this.status = PipelineStatus.NotStarted;
 	}
 
-	private PipelineStatus _start() {
-		// The following pipeline process could
-		// be refactored into something more ergonomic and elegant.
-		// This will do for now.
-
-		var status = pull();
-		if (status != PipelineStatus.Ok) {
-			return status;
-		}
+	/**
+	 * Add components to the Pipeline.
+	 * 
+	 * @param components the components to add
+	 * @return the pipeline
+	 */
+	public Pipeline withComponent(PipelineComponent... components) {
+		this.components.addAll(Arrays.asList(components));
 
-		status = lint();
-		if (status != PipelineStatus.Ok) {
-			return status;
-		}
+		return this;
+	}
 
-		status = compile();
-		if (status != PipelineStatus.Ok) {
-			return status;
+	private PipelineStatus _start() {
+		for (var component : components) {
+			var status = component.execute(pipelineDir, event);
+			if (status != PipelineStatus.Ok) {
+				return status;
+			}
 		}
 
 		return PipelineStatus.Ok;
 	}
 
 	/**
-	 * Start the pipeline.
+	 * Start the pipeline with the added components.
 	 * 
-	 * @return the status of the executed pipeline. OK if everything went ok.
+	 * @return the status of the pipeline execution
 	 */
 	public PipelineStatus start() {
 		this.status = PipelineStatus.InProgress;
 		this.status = _start();
 
 		return this.status;
 	}
```

Comment:
What's the point of the `_start()` method? It seems like a very intentional decision, but I can't see the reason for it in this case. My suggestion would be to combine them into one.

```java
	/**
	 * Start the pipeline with the added components.
	 *
	 * @return the status of the pipeline execution
	 */
	public PipelineStatus start() {
		for (var component : components) {
			var status = component.execute(pipelineDir, event);
			if (status != PipelineStatus.Ok) {
				return status;
			}
		}

		return PipelineStatus.Ok;
	}
```

Reply:
You're right, it was used before but not anymore.